taichi/transforms/constant_fold.cpp

```diff
@@ -42,30 +47,86 @@ class ConstantFold : public BasicStmtVisitor {
         throw IRModified();
       }
     }
+#endif
+  }
+
+  struct BinaryEvaluatorId
+  {
+    BinaryOpType op;
+    DataType ret, lhs, rhs;
+
+    explicit operator int() const // make STL map happy
+    {
+      return (int)op | (int) ret << 8 | (int) lhs << 16 | (int) rhs << 24;
+    }
+  };
+
+  static Kernel *get_binary_op_jit_eval_kernel(BinaryEvaluatorId const &id)
+  {
+    auto &cache = get_current_program().jit_evaluator_cache;
+#if 1
+    int iid = int(id);
+    auto it = cache.find(iid);
+    if (it != cache.end()) // cached?
+      return it->second.get();
+#endif
+    static int jic = 0; // X: race?
+    auto kernel_name = fmt::format("jit_evaluator_{}", jic++);
+    auto func = [&] () {
+      auto lhstmt = Stmt::make<ArgLoadStmt>(1, false);
+      auto rhstmt = Stmt::make<ArgLoadStmt>(2, false);
+      auto oper = Stmt::make<BinaryOpStmt>(id.op, lhstmt.get(), rhstmt.get());
+      auto ret = Stmt::make<ArgStoreStmt>(0, oper.get());
+      current_ast_builder().insert(std::move(lhstmt));
+      current_ast_builder().insert(std::move(rhstmt));
+      current_ast_builder().insert(std::move(oper));
+      current_ast_builder().insert(std::move(ret));
+    };
+    auto ker = std::make_unique<Kernel>(get_current_program(), func, kernel_name);
+    ker->insert_arg(id.ret, false);
+    ker->insert_arg(id.lhs, false);
+    ker->insert_arg(id.rhs, false);
+    ker->mark_arg_return_value(0, true);
+    auto *ker_ptr = ker.get();
+#if 1
+    cache[iid] = std::move(ker);
+#endif
+    return ker_ptr;
+  }
+
+  static bool jit_from_binary_op(TypedConstant &ret, BinaryOpType op,
+      const TypedConstant &lhs, const TypedConstant &rhs)
+  {
+    BinaryEvaluatorId id{op, ret.dt, lhs.dt, rhs.dt};
+    auto *ker = get_binary_op_jit_eval_kernel(id);
+    auto &ctx = get_current_program().context;
+#define PER_TYPE(x) ctx.set_arg(1, lhs.val_##x);
+    PER_TYPE(i32) PER_TYPE(i64)
+    PER_TYPE(f32) PER_TYPE(f64)
+#undef PER_TYPE
+#define PER_TYPE(x) ctx.set_arg(2, rhs.val_##x);
+    PER_TYPE(i32) PER_TYPE(i64)
+    PER_TYPE(f32) PER_TYPE(f64)
+#undef PER_TYPE
```

Comment:
I would suggest extracting it into a function like `void set_arg(const TypedConstant&)` and avoid the duplicated code.

In that function, I think it would be clearer to avoid using macros. Writing `if` 4 times for 4 types is fine.